Add APIs to get JVM names of functions and property accessors

yigit opened the pull request:
This PR adds a new API `getJvmName` to get the jvm name of a method or accessor.

This is crucial for annotation processors that generate java source code as they might need to access internal methods in the same module compilation (as it works w/ KAPT).

Fixes #129

ting-yuan commented on api/src/main/kotlin/com/google/devtools/ksp/processing/Resolver.kt:
```diff
@@ -141,4 +141,33 @@ interface Resolver {
+     * Returns the jvm name of the given function.
+     *
+     * Note that this might be different from the name declared in the Kotlin source code in two cases:
+     * a) If the function receives or returns an internal class, its name will be mangled according to
```
s/internal/inline

ting-yuan commented on api/src/main/kotlin/com/google/devtools/ksp/processing/Resolver.kt:
```diff
@@ -141,4 +141,33 @@ interface Resolver {
     ): KSFunction
+
+    /**
+     * Returns the jvm name of the given function.
```
Can you also mention that this ties to Kotlin compiler's implementation and is not guaranteed to be cross-version compatible? Namely, not guaranteed to work on libraries compiled by other version of compilers.

yigit replied:
updated. lmk.

ting-yuan commented on api/src/main/kotlin/com/google/devtools/ksp/processing/Resolver.kt:
```diff
@@ -141,4 +141,33 @@ interface Resolver {
+     * By default, this name will match the name calculated according to
+     * https://kotlinlang.org/docs/reference/java-to-kotlin-interop.html#properties.
+     * Note that the result of this function might be different from that name in two cases:
+     * a) If the property's type is an internal class, accessor's name will be mangled according to
```
s/internal/inline

ting-yuan commented on compiler-plugin/testData/api/mangledNames.kt:
```diff
@@ -0,0 +1,94 @@
+// internalInlineReturningFun -> internalInlineReturningFun-HRn7Rpw$mainModule
+// fileLevelInternalFun -> fileLevelInternalFun
+// fileLevelInlineReceivingFun -> fileLevelInlineReceivingFun-E03SJzc
+// fileLevelInlineReturningFun -> fileLevelInlineReturningFun
```
Just curious, why this and `fileLevelInternalInlineReturningFun`  isn't mangled?

yigit replied:
i honestly don't know 🤦 
I got these outputs from compiling the same input on a sample project and it is not mangled there.

```
➜  yigit-google-ksp git:(mangled-names) javap -p ~/src/ManglingPlayground/mainModule/build/classes/kotlin/main/mainPackage/InputKt.class
Compiled from "input.kt"
public final class mainPackage.InputKt {
  public static final void fileLevelInternalFun();
  public static final void fileLevelInlineReceivingFun-E03SJzc(java.lang.String);
  public static final java.lang.String fileLevelInlineReturningFun();
  public static final void fileLevelInternalInlineReceivingFun-E03SJzc(java.lang.String);
  public static final java.lang.String fileLevelInternalInlineReturningFun();
}
```
vs it is mangled when it is on a class:
```
➜  yigit-google-ksp git:(mangled-names) javap -p ~/src/ManglingPlayground/mainModule/build/classes/kotlin/main/mainPackage/Foo.class    
Compiled from "input.kt"
public final class mainPackage.Foo {
  private java.lang.String normalProp;
  private java.lang.String inlineProp;
  private java.lang.String internalProp;
  private java.lang.String internalInlineProp;
  public final java.lang.String getNormalProp();
  public final void setNormalProp(java.lang.String);
  public final java.lang.String getInlineProp-HRn7Rpw();
  public final void setInlineProp-E03SJzc(java.lang.String);
  public final java.lang.String getInternalProp$mainModule();
  public final void setInternalProp$mainModule(java.lang.String);
  public final java.lang.String getInternalInlineProp-HRn7Rpw$mainModule();
  public final void setInternalInlineProp-E03SJzc$mainModule(java.lang.String);
  public final void normalFun();
  public final void inlineReceivingFun-E03SJzc(java.lang.String);
  public final java.lang.String inlineReturningFun-HRn7Rpw();
  public final void internalInlineReceivingFun-E03SJzc$mainModule(java.lang.String);
  public final java.lang.String internalInlineReturningFun-HRn7Rpw$mainModule();
  public mainPackage.Foo();
}
```

yigit commented:
addressed the review comments and also added a test case with JvmName. Which is also another bonus of this CL as it was something that Room would need to handle manually.

yigit commented:
can we squash & merge this one? (I don't have merge access)

yigit commented:
added KspExperimental annotation.
btw, also noticed the comment above was for another PR :facepalm:
